go/arrow/array/booleanbuilder.go

```diff
@@ -40,17 +37,12 @@ func NewBooleanBuilder(mem memory.Allocator) *BooleanBuilder {
 // When the reference count goes to zero, the memory is freed.
 // Release may be called simultaneously from multiple goroutines.
 func (b *BooleanBuilder) Release() {
-	debug.Assert(atomic.LoadInt64(&b.refCount) > 0, "too many releases")
-
-	if atomic.AddInt64(&b.refCount, -1) == 0 {
-		if b.nullBitmap != nil {
-			b.nullBitmap.Release()
-			b.nullBitmap = nil
-		}
-		if b.data != nil {
-			b.data.Release()
-			b.data = nil
-		}
+	b.builder.Release()
+
+	if b.data != nil {
```

Comment:
wouldn't this mean only the `atomic` block from inside `b.builder.Release` is guaranteed to see only one goroutine, but not the `if` block conditioned by `b.builder.Release`?

(perhaps best to leave this code alone, actually?)

Reply:
I agree, it may be better just to leave the code, however, if the `builder.Release` method returns the new reference count:

```go
 // Release decreases the reference count by 1.
 // When the reference count goes to zero, the memory is freed.
 func (b *builder) Release() int64 {
 	debug.Assert(atomic.LoadInt64(&b.refCount) > 0, "too many releases")

 	newref := atomic.AddInt64(&b.refCount, -1)
 	if newref == 0 {
 		if b.nullBitmap != nil {
 			b.nullBitmap.Release()
 			b.nullBitmap = nil
 		}
 	}
 	return newref
 }
```

it is the equivalent of

```go
if atomic.AddInt64(&b.refCount -1) == 0 {
    b.data != nil {
```